# 데이터 로드

In [1]:
%config Completer.use_jedi = False
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from datetime import datetime
from datetime import datetime, timedelta
from dateutil.relativedelta import relativedelta

%matplotlib inline

In [2]:
df = pd.read_excel("total.xlsx")
print(df.shape)
df.head()

(7043, 46)


Customer ID  Count  Gender  Age Under 30 Senior Citizen Married Dependents  \
0  8779-QRDMV      1    Male   78       No            Yes      No         No   
1  7495-OOKFY      1  Female   74       No            Yes     Yes        Yes   
2  1658-BYGOY      1    Male   71       No            Yes      No        Yes   
3  4598-XLKNJ      1  Female   78       No            Yes     Yes        Yes   
4  4846-WHAFZ      1  Female   80       No            Yes     Yes        Yes   

   Number of Dependents     Service ID  ... Total Revenue   Status ID  \
0                     0  IJKDQVSWH3522  ...         59.65  SWSORB1252   
1                     1  BFKMZJAIE2285  ...       1024.10  SNAEQA8572   
2                     3  EIMVJQBMT7187  ...       1910.88  LMBQNN3714   
3                     1  EROZQXDUU4979  ...       2995.07  VRZYZI9978   
4                     1  GEEYSJUHY6991  ...       3102.36  FDNAKX1688   

   Satisfaction Score Customer Status Churn Label  Churn Value Churn Score  \
0                   3         Churned         Yes            1          91   
1                   3         Churned         Yes            1          69   
2                   2         Churned         Yes            1          81   
3                   2         Churned         Yes            1          88   
4                   2         Churned         Yes            1          67   

   CLTV   Churn Category                  Churn Reason  
0  5433       Competitor  Competitor offered more data  
1  5302       Competitor  Competitor made better offer  
2  3179       Competitor  Competitor made better offer  
3  5337  Dissatisfaction     Limited range of services  
4  2793            Price            Extra data charges  

[5 rows x 46 columns]

In [3]:
# Customer ID & 선정한 컬럼 & 'Churn Label', 'Churn Score'
selected_cols = ['Customer ID','Dependents', 'Under 30', 'Monthly Charge', 'Total Revenue', 'Tenure in Months', 'Internet Service', 
                 'Device Protection Plan', 'Unlimited Data', 'Streaming Movies', 'Churn Label', 'Churn Score']

new_df = df[selected_cols]

new_df

Customer ID Dependents Under 30  Monthly Charge  Total Revenue  \
0     8779-QRDMV         No       No           39.65          59.65   
1     7495-OOKFY        Yes       No           80.65        1024.10   
2     1658-BYGOY        Yes       No           95.45        1910.88   
3     4598-XLKNJ        Yes       No           98.50        2995.07   
4     4846-WHAFZ        Yes       No           76.50        3102.36   
...          ...        ...      ...             ...            ...   
7038  2569-WGERO         No       No           21.15        3039.53   
7039  6840-RESVB        Yes       No           84.80        2807.47   
7040  2234-XADUH        Yes       No          103.20        9453.04   
7041  4801-JZAZL        Yes       No           29.60         319.21   
7042  3186-AJIEK         No       No          105.65        8887.86   

      Tenure in Months Internet Service Device Protection Plan Unlimited Data  \
0                    1              Yes                    Yes             No   
1                    8              Yes                     No            Yes   
2                   18              Yes                     No            Yes   
3                   25              Yes                    Yes            Yes   
4                   37              Yes                     No            Yes   
...                ...              ...                    ...            ...   
7038                72               No                     No             No   
7039                24              Yes                    Yes            Yes   
7040                72              Yes                    Yes            Yes   
7041                11              Yes                     No            Yes   
7042                66              Yes                    Yes            Yes   

     Streaming Movies Churn Label  Churn Score  
0                 Yes         Yes           91  
1                  No         Yes           69  
2                 Yes         Yes           81  
3                 Yes         Yes           88  
4                  No         Yes           67  
...               ...         ...          ...  
7038               No          No           45  
7039              Yes          No           59  
7040              Yes          No           71  
7041               No          No           59  
7042              Yes          No           38  

[7043 rows x 12 columns]

# label 값 맞추기

## Yes, No -> 1, 0

In [5]:
target_cols = ['Under 30','Dependents', 'Internet Service','Device Protection Plan','Streaming Movies', 'Unlimited Data','Churn Label']

for col in target_cols:
    new_df[col] = new_df[col].map({'Yes': 1, 'No': 0})
    

new_df.head()

C:\Users\ksh\AppData\Local\Temp\ipykernel_16008\2924241732.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df[col] = new_df[col].map({'Yes': 1, 'No': 0})
C:\Users\ksh\AppData\Local\Temp\ipykernel_16008\2924241732.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df[col] = new_df[col].map({'Yes': 1, 'No': 0})
C:\Users\ksh\AppData\Local\Temp\ipykernel_16008\2924241732.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_

Customer ID  Dependents  Under 30  Monthly Charge  Total Revenue  \
0  8779-QRDMV           0         0           39.65          59.65   
1  7495-OOKFY           1         0           80.65        1024.10   
2  1658-BYGOY           1         0           95.45        1910.88   
3  4598-XLKNJ           1         0           98.50        2995.07   
4  4846-WHAFZ           1         0           76.50        3102.36   

   Tenure in Months  Internet Service  Device Protection Plan  Unlimited Data  \
0                 1                 1                       1               0   
1                 8                 1                       0               1   
2                18                 1                       0               1   
3                25                 1                       1               1   
4                37                 1                       0               1   

   Streaming Movies  Churn Label  Churn Score  
0                 1            1           91  
1                 0            1           69  
2                 1            1           81  
3                 1            1           88  
4                 0            1           67

# 정규화

In [6]:
from sklearn.cluster import KMeans
from sklearn.preprocessing import MinMaxScaler

scaled_df = new_df.copy()

# MinMaxScaler를 사용한 정규화
scaler = MinMaxScaler()
scaled_df[['Tenure in Months','Monthly Charge', 'Total Revenue']] = scaler.fit_transform(scaled_df[['Tenure in Months','Monthly Charge', 'Total Revenue']])

In [7]:
q1 = scaled_df['Tenure in Months'].quantile(0.25)
q1

0.11267605633802817

In [8]:
scaled_df

Customer ID  Dependents  Under 30  Monthly Charge  Total Revenue  \
0     8779-QRDMV           0         0        0.212935       0.003202   
1     7495-OOKFY           1         0        0.620896       0.083855   
2     1658-BYGOY           1         0        0.768159       0.158013   
3     4598-XLKNJ           1         0        0.798507       0.248680   
4     4846-WHAFZ           1         0        0.579602       0.257652   
...          ...         ...       ...             ...            ...   
7038  2569-WGERO           0         0        0.028856       0.252398   
7039  6840-RESVB           1         0        0.662189       0.232992   
7040  2234-XADUH           1         0        0.845274       0.788735   
7041  4801-JZAZL           1         0        0.112935       0.024908   
7042  3186-AJIEK           0         0        0.869652       0.741471   

      Tenure in Months  Internet Service  Device Protection Plan  \
0             0.000000                 1                       1   
1             0.098592                 1                       0   
2             0.239437                 1                       0   
3             0.338028                 1                       1   
4             0.507042                 1                       0   
...                ...               ...                     ...   
7038          1.000000                 0                       0   
7039          0.323944                 1                       1   
7040          1.000000                 1                       1   
7041          0.140845                 1                       0   
7042          0.915493                 1                       1   

      Unlimited Data  Streaming Movies  Churn Label  Churn Score  
0                  0                 1            1           91  
1                  1                 0            1           69  
2                  1                 1            1           81  
3                  1                 1            1           88  
4                  1                 0            1           67  
...              ...               ...          ...          ...  
7038               0                 0            0           45  
7039               1                 1            0           59  
7040               1                 1            0           71  
7041               1                 0            0           59  
7042               1                 1            0           38  

[7043 rows x 12 columns]

# 학습과 예측에 사용할 데이터셋 만들기

In [9]:
X = scaled_df.iloc[:,1:-2]
print(X.shape)
X

(7043, 9)


Dependents  Under 30  Monthly Charge  Total Revenue  Tenure in Months  \
0              0         0        0.212935       0.003202          0.000000   
1              1         0        0.620896       0.083855          0.098592   
2              1         0        0.768159       0.158013          0.239437   
3              1         0        0.798507       0.248680          0.338028   
4              1         0        0.579602       0.257652          0.507042   
...          ...       ...             ...            ...               ...   
7038           0         0        0.028856       0.252398          1.000000   
7039           1         0        0.662189       0.232992          0.323944   
7040           1         0        0.845274       0.788735          1.000000   
7041           1         0        0.112935       0.024908          0.140845   
7042           0         0        0.869652       0.741471          0.915493   

      Internet Service  Device Protection Plan  Unlimited Data  \
0                    1                       1               0   
1                    1                       0               1   
2                    1                       0               1   
3                    1                       1               1   
4                    1                       0               1   
...                ...                     ...             ...   
7038                 0                       0               0   
7039                 1                       1               1   
7040                 1                       1               1   
7041                 1                       0               1   
7042                 1                       1               1   

      Streaming Movies  
0                    1  
1                    0  
2                    1  
3                    1  
4                    0  
...                ...  
7038                 0  
7039                 1  
7040                 1  
7041                 0  
7042                 1  

[7043 rows x 9 columns]

In [11]:
y = scaled_df["Churn Label"]
print(y.shape)
y

(7043,)


0       1
1       1
2       1
3       1
4       1
       ..
7038    0
7039    0
7040    0
7041    0
7042    0
Name: Churn Label, Length: 7043, dtype: int64

In [12]:
# 사이킷런에서 제공하는 model_selection의 train_test_split으로 만들기
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
                                    X, y, test_size=0.2, random_state=42)

In [13]:
# train 세트의 데이터 수를 확인
X_train.shape, y_train.shape

((5634, 9), (5634,))

In [14]:
# test 세트의 데이터 수를 확인
X_test.shape, y_test.shape

((1409, 9), (1409,))

# 알고리즘

## 로지스틱회귀

In [19]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(random_state=42)
model.fit(X_train, y_train)

LogisticRegression(random_state=42)

In [20]:
y_predict = model.predict(X_test)
y_predict

array([0, 0, 0, ..., 1, 0, 0], dtype=int64)

In [21]:
diff_count = (y_predict != y_test).sum() 
diff_count

316

In [22]:
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(y_test, y_predict)
print("Accuracy:", accuracy)

Accuracy: 0.7757274662881476


## DecisionTree

In [24]:
from sklearn.model_selection import GridSearchCV

dt_clf = DecisionTreeClassifier(random_state=42)

parameters = {'max_depth':[2, 3, 5, 10],
             'min_samples_split':[2, 3, 5],
             'min_samples_leaf':[1, 5, 8]}

# DecisionTreeClassifier 객체 생성 후 GridSearchCV 수행
dt_grid_cv = GridSearchCV(dt_clf , param_grid=parameters , cv=2, n_jobs=-1 )
dt_grid_cv.fit(X_train , y_train)

print('최적 하이퍼 파라미터:\n', dt_grid_cv.best_params_)
print('최고 예측 정확도: {0:.4f}'.format(dt_grid_cv.best_score_))

최적 하이퍼 파라미터:
 {'max_depth': 5, 'min_samples_leaf': 8, 'min_samples_split': 2}
최고 예측 정확도: 0.7902


## RandomForest

In [25]:
from sklearn.model_selection import GridSearchCV

rf_clf = RandomForestClassifier(random_state=42)

params = {
    'n_estimators':[100], # 결정 트리 개수 지정
    'max_depth' : [6, 8, 10, 12], # 트리의 최대 깊이 지정
    'min_samples_leaf' : [8, 12, 18 ], # 분할된 경우 왼족, 오른쪽의 브랜치 노드에서 가져가야 할 최소한의 샘플 데이터 수
    'min_samples_split' : [8, 16, 20] # 노드를 분할하기 위한 최소한의 샘플 데이터 수 => 과적합 제어
}
# RandomForestClassifier 객체 생성 후 GridSearchCV 수행
rf_grid_cv = GridSearchCV(rf_clf , param_grid=params , cv=2, n_jobs=-1 )
rf_grid_cv.fit(X_train , y_train)

print('최적 하이퍼 파라미터:\n', rf_grid_cv.best_params_)
print('최고 예측 정확도: {0:.4f}'.format(rf_grid_cv.best_score_))

최적 하이퍼 파라미터:
 {'max_depth': 10, 'min_samples_leaf': 8, 'min_samples_split': 8, 'n_estimators': 100}
최고 예측 정확도: 0.8012


## GradientBoosting

In [26]:
from sklearn.model_selection import GridSearchCV

gb_clf = GradientBoostingClassifier(random_state=42)

params = {
    'n_estimators':[100], # 결정 트리 개수 지정
    'max_depth' : [6, 8, 10, 12], # 트리의 최대 깊이 지정
    'min_samples_leaf' : [8, 12, 18,20,24 ], # 분할된 경우 왼족, 오른쪽의 브랜치 노드에서 가져가야 할 최소한의 샘플 데이터 수
    'min_samples_split' : [8, 16, 20,22,26] # 노드를 분할하기 위한 최소한의 샘플 데이터 수 => 과적합 제어
}


gb_grid_cv = GridSearchCV(gb_clf , param_grid=params , cv=2, n_jobs=-1 )
gb_grid_cv.fit(X_train , y_train)

print('최적 하이퍼 파라미터:\n', gb_grid_cv.best_params_)
print('최고 예측 정확도: {0:.4f}'.format(gb_grid_cv.best_score_))

최적 하이퍼 파라미터:
 {'max_depth': 6, 'min_samples_leaf': 20, 'min_samples_split': 8, 'n_estimators': 100}
최고 예측 정확도: 0.7895


## XGBoost

In [27]:
from sklearn.model_selection import GridSearchCV

XG_clf = XGBClassifier(random_state=101)

params = {'n_estimators':[100,200,300,400], 'learning_rate' : [0.1, 0.2, 0.3, 0.4], 'max_depth' : [1,2,3,4,5]}

xg_grid_cv = GridSearchCV(XG_clf , param_grid=params , cv=2, n_jobs=-1 )
xg_grid_cv.fit(X_train , y_train)

print('최적 하이퍼 파라미터:\n', xg_grid_cv.best_params_)
print('최고 예측 정확도: {0:.4f}'.format(xg_grid_cv.best_score_))

최적 하이퍼 파라미터:
 {'learning_rate': 0.1, 'max_depth': 1, 'n_estimators': 300}
최고 예측 정확도: 0.8014


# 알고리즘 선택

target_cols = ['Under 30','Dependents', 'Internet Service','Device Protection Plan','Streaming Movies', 'Unlimited Data']


* DecisionTree : 0.7902	
* RandomForest : 0.8012
* GradientBoosting : 0.7895
* LogisticRegression : 0.7757274662881476
* **XGBoost : 0.8014**

# 가상 고객 테스트

In [28]:
from sklearn.model_selection import GridSearchCV

XG_clf = XGBClassifier(random_state=101)

params = {'n_estimators':[100,200,300,400], 'learning_rate' : [0.1, 0.2, 0.3, 0.4], 'max_depth' : [1,2,3,4,5]}

xg_grid_cv = GridSearchCV(XG_clf , param_grid=params , cv=2, n_jobs=-1 )
xg_grid_cv.fit(X_train , y_train)

print('최적 하이퍼 파라미터:\n', xg_grid_cv.best_params_)
print('최고 예측 정확도: {0:.4f}'.format(xg_grid_cv.best_score_))

최적 하이퍼 파라미터:
 {'learning_rate': 0.1, 'max_depth': 1, 'n_estimators': 300}
최고 예측 정확도: 0.8014


In [46]:
# Dependents, Under 30, Monthly Charge, Total Revenue, Tenure in Months	Internet Service, Device Protection Plan, 
# Unlimited Data, Streaming Movies

# 최적의 모델을 사용하여 테스트 데이터를 예측합니다.
test_data = [[0, 1, 0.8628, 0.7, 0.11267, 1, 1, 0, 1]]  # 테스트 데이터를 리스트로 준비
predicted_class = xg_grid_cv.best_estimator_.predict(test_data)

# 최적의 모델을 사용하여 테스트 데이터의 클래스 확률을 예측합니다.
predicted_proba = xg_grid_cv.best_estimator_.predict_proba(test_data)

# 클래스 1일 확률(이탈할 확률)을 출력합니다.
probability_of_class_1 = predicted_proba[0][1]  # 첫 번째 테스트 데이터에 대한 클래스 1일 확률
print("가상 고객이 이탈할 확률:", probability_of_class_1)
print("가상 고객 이탈 점수:", int(probability_of_class_1 * 100), "점")

가상 고객이 이탈할 확률: 0.644448
가상 고객 이탈 점수: 64 점


# 고객 그룹화

이탈 점수(이탈 위험도)에 따라 기존 고객들 그룹화 & 새로운 가상 고객 그룹이 어딘지 정보 제공

* High : 0 ~ 25점
* Mid : 26 ~ 50점
* Low : 51 ~ 75점
* Stay : 76 ~ 100점